src/halmos/__main__.py

```diff
@@ -571,8 +592,8 @@ def step_invariant_tests(
                 visited.add(post_id)
 
                 # update call traces
-                post_ex.context = deepcopy(pre_ex.context)
-                post_ex.context.trace.append(subcall)
+                post_ex.context = copy(pre_ex.context)
```

Comment:
isn't the trace of this shallow-copied context supposed to be updated when pre_ex.trace is extended later? or, do we not need to worry because pre_ex.trace will never be extended?

Reply:
well tbh deepcopy didn't feel right, but it's also not entirely clear to me why we should copy at all. Semantically, that doesn't make a lot of sense to me. The trace shouldn't be extended, that would be weird to me (but the call sequence should be!)

Can you check again with d329b7a? No copy at all, turns out the only part we need from the context is the immutable Message from setup_ex.

Reply:
(eventually I would like to have a separate NetworkState field in Exec, so that we can explicitly pluck out the state from an exec and discard intermediate values like the EVMState and CallContext)

Reply:
> Can you check again with https://github.com/a16z/halmos/commit/d329b7a25d814c7afdca9a628132f4ab09fe9211? No copy at all, turns out the only part we need from the context is the immutable Message from setup_ex.

looks good to me, thank you!